Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp

```diff
@@ -0,0 +1,250 @@
+    : m_numerator(move(numerator))
+    , m_denominator(move(denominator))
+{
+    VERIFY(m_denominator != 0);
+    reduce();
+}
+
+BigFraction::BigFraction(Crypto::SignedBigInteger value)
+    : BigFraction(move(value), 1)
+{
+}
+
+BigFraction::BigFraction(StringView sv)
+{
+    auto maybe_dot_index = sv.find('.');
+
+    auto integer_part_view = sv.substring_view(0, maybe_dot_index.value_or(sv.length()));
+    auto fraction_part_view = maybe_dot_index.has_value() ? sv.substring_view(1 + *maybe_dot_index) : "0"sv;
+
+    auto integer_part = Crypto::SignedBigInteger::from_base(10, integer_part_view);
+    auto fractional_part = Crypto::SignedBigInteger::from_base(10, fraction_part_view);
+    auto fraction_length = Crypto::UnsignedBigInteger::create_from(fraction_part_view.length());
+
+    *this = BigFraction(move(integer_part)) + BigFraction(move(fractional_part), Crypto::NumberTheory::Power("10"_bigint, move(fraction_length)));
+};
+
+BigFraction BigFraction::operator+(BigFraction const& rhs) const
+{
+    if (rhs.m_numerator == "0"_bigint)
+        return *this;
+
+    auto tmp = *this;
+    tmp.m_numerator.set_to(m_numerator.multiplied_by(rhs.m_denominator).plus(rhs.m_numerator.multiplied_by(m_denominator)));
+    tmp.m_denominator.set_to(m_denominator.multiplied_by(rhs.m_denominator));
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+BigFraction BigFraction::operator-(BigFraction const& rhs) const
+{
+    return *this + (-rhs);
+}
+
+BigFraction BigFraction::operator*(BigFraction const& rhs) const
+{
+    auto tmp = *this;
+    tmp.m_numerator.set_to(tmp.m_numerator.multiplied_by(rhs.m_numerator));
+    tmp.m_denominator.set_to(tmp.m_denominator.multiplied_by(rhs.m_denominator));
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+BigFraction BigFraction::operator-() const
+{
+    return { m_numerator.negated_value(), m_denominator };
+}
+
+BigFraction BigFraction::invert() const
+{
+    return BigFraction { 1 } / *this;
+}
+
+BigFraction BigFraction::operator/(BigFraction const& rhs) const
+{
+    VERIFY(rhs.m_numerator != "0"_bigint);
+
+    auto tmp = *this;
+    tmp.m_numerator.set_to(m_numerator.multiplied_by(rhs.m_denominator));
+    tmp.m_denominator.set_to(m_denominator.multiplied_by(rhs.m_numerator.unsigned_value()));
+
+    if (rhs.m_numerator.is_negative())
+        tmp.m_numerator.negate();
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+bool BigFraction::operator<(BigFraction const& rhs) const
+{
+    return (*this - rhs).m_numerator.is_negative();
+}
+
+bool BigFraction::operator==(BigFraction const& rhs) const
+{
+    return m_numerator == rhs.m_numerator && m_denominator == rhs.m_denominator;
+}
+
+BigFraction::BigFraction(double d)
+{
+    bool negative = false;
+    if (d < 0) {
+        negative = true;
+        d = -d;
+    }
+    i8 current_pow = 0;
+    while (AK::pow(10.0, (double)current_pow) <= d)
+        current_pow += 1;
+    current_pow -= 1;
+    double epsilon = 1e-6;
```

Comment:
Where does this come from? Shouldn't it be constexpr? Why isn't it 10^-16?

Reply:
To answer your question, it comes from `KeypadValue::KeypadValue(double d)` and I didn't bother spending time on it.
It can definitely be constexpr and for the value I think it is related to the other magic number `m_decimal_places > 6`.
And 6 was the number of decimal supported by the calculator.

We can definitely discuss the best value for it, but this code shouldn't last long. It is only used in `BigFraction::inexact_sqrt` and I already have something to get rid of the usage of doubles.

Reply:
Okay, because this is a library function it shouldn't assume the user is Calculator who needs 6 decimal places max (which might not be enough either way...). I'm thinking of 10^-16 because that's the approximate decimal precision for doubles close to zero. On the other hand, there should be a library function (or builtin) providing the `eps` constant as specified by IEEE 754, which should be the exact value we want to use.

Reply:
It was easier than I thought, we already implement the value with `NumericLimits<double>::epsilon()`.
I also removed the `break` when `m_decimal_places > 6`